src/bin/ipa_bench/models.rs

```diff
@@ -109,57 +111,169 @@ impl SecretSharable for u64 {
 
-    /// A key to group sets of the events.
-    pub breakdown_key: Number,
+    /// An epoch in which this event is generated. Using an 8-bit value = 256 epochs > 4 years (assuming 1 epoch = 1 week).
+    ///
+    /// How we send an epoch epoch value between Report Collector and Helper network is still an open question. We will
+    /// either send it in the clear, or mix it in MAC. For now, we can assume that this struct is a MPC internal data
+    /// model.
+    #[allow(clippy::cast_possible_truncation)]
+    fn epoch(&self) -> Self::Epoch {
+        (self.0 / Self::SECONDS_IN_EPOCH) as Self::Epoch
+    }
+
+    /// An offset in seconds into a given epoch. Max value is `SECONDS_IN_EPOCH - 1`.
+    ///
+    /// Use `u32` (`< 2^20 seconds`) to leverage simple arithmetics, but we drop the first 12 bits when serializing.
+    /// That'll save us ~1.5G with 1B rows.
+    fn offset(&self) -> Self::Offset {
+        (self.0 % Self::SECONDS_IN_EPOCH) as Self::Offset
+    }
 }
 
-#[cfg(feature = "debug")]
-impl Debug for SourceEvent {
-    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
-        write!(
-            f,
-            "SourceEvent:\n  matchkeys={:?}\n  breakdown_key={}",
-            &self.event.matchkeys, &self.breakdown_key
-        )
+impl From<u64> for EventTimestamp {
+    fn from(v: u64) -> Self {
+        EventTimestamp(v)
     }
 }
 
-#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
-pub struct TriggerEvent {
-    pub event: Event,
+impl Add for EventTimestamp {
+    type Output = Self;
 
-    /// Conversion value.
-    pub value: SecretShare,
+    fn add(self, rhs: Self) -> Self::Output {
+        EventTimestamp(self.0 + rhs.0)
+    }
 }
 
-#[cfg(feature = "debug")]
-impl Debug for TriggerEvent {
-    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
-        write!(
-            f,
-            "TriggerEvent:\n  matchkeys={:?}\n  value={:?}",
-            &self.event.matchkeys, &self.value
-        )
+impl AddAssign for EventTimestamp {
+    // #[allow(clippy::assign_op_pattern)]
+    fn add_assign(&mut self, rhs: Self) {
+        *self = *self + rhs;
+    }
+}
+
+impl PartialOrd for EventTimestamp {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(u64::from(*self).cmp(&u64::from(*other)))
+    }
+}
+
+impl From<EventTimestamp> for u64 {
+    fn from(v: EventTimestamp) -> Self {
+        v.0 % (EventTimestamp::MAX + 1)
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct Event {
+    // An identifier, set in the user agent, which identifies an individual person. This must never be released (beyond
+    /// the match key provider) to any party in unencrypted form. For the purpose of this tool, however, the value is in
+    /// clear and not secret shared.
+    pub matchkey: MatchKey,
+
+    /// An identifier, specified by the report collector, to denote if a given pair of source and trigger events can be
```

Comment:
huh, what is this identifier for? specifically, in what scenarios would it be `None`, and what is the significance of the `Number` when it is there?

Reply:
This is in our [e2e doc](https://github.com/patcg-individual-drafts/ipa/blob/main/IPA-End-to-End.md#types-of-queries), which is used together with match keys for attribution.

We only know that it will be an `int`. This identifier is also optional (only use match keys in attribution process). Using `Option` complicates the logic because we have to deal with `None` but I couldn't use `0` to indicate that the value is non-existent since `0` is a valid integer.